File: nRF24L01p.py
#!/usr/bin/python
# 
# Library for ISP communication with nRF24L01+ for raspberrypi.
# Uses spidev: https://github.com/doceme/py-spidev
# For a higher level library, see https://github.com/jpbarraca/pynrf24/blob/master/nrf24.py

import time
import spidev
import RPi.GPIO as gpio

class nRF24L01p:

    # Instruction Mnemonics
    R_REGISTER = 0x00
    W_REGISTER = 0x20
    R_RX_PAYLOAD = 0x61
    W_TX_PAYLOAD = 0xA0
    FLUSH_TX = 0xE1
    FLUSH_RX = 0xE2
    REUSE_TX_PL = 0xE3
    R_RX_PL_WID = 0x60
    W_ACK_PAYLOAD = 0xA8
    W_TX_PAYLOAD_NOACK = 0xB0
    NOP = 0xFF

    # Register addresses
    CONFIG = 0x00
    EN_AA = 0x01
    EN_RXADDR = 0x02
    SETUP_AW = 0x03
    SETUP_RETR = 0x04
    RF_CH = 0x05
    RF_SETUP = 0x06
    STATUS = 0x07
    OBSERVE_TX = 0x08
    RPD = 0x09
    RX_ADDR_P0 = 0x0A
    RX_ADDR_P1 = 0x0B
    RX_ADDR_P2 = 0x0C
    RX_ADDR_P3 = 0x0D
    RX_ADDR_P4 = 0x0E
    RX_ADDR_P5 = 0x0F
    TX_ADDR = 0x10
    RX_PW_P0 = 0x11
    RX_PW_P1 = 0x12
    RX_PW_P2 = 0x13
    RX_PW_P3 = 0x14
    RX_PW_P4 = 0x15
    RX_PW_P5 = 0x16
    FIFO_STATUS = 0x17
    DYNPD = 0x1C
    FEATURE = 0x1D

    def __init__(self, spi_device=0, CE_port=24, IRQ_port=25):
        """Initialize spi communication with device. IRQ and CE pins are set in BCM mode.
            Note: CE0 and CE1 are selected via spi_device, and connect to CSN."""
        self.spi = spidev.SpiDev()
        self.spi.open( 0, 0)
        self.spi.bits_per_word = 8
        self.spi.max_speed_hz = 10000000
        self.CE = CE_port
        self.IRQ = IRQ_port
        gpio.setmode(gpio.BCM)
        gpio.setup( self.CE, gpio.OUT, initial=gpio.LOW)
        gpio.setup( self.IRQ, gpio.IN)

    def __del__(self):
        self.spi.close()
        gpio.cleanup()

    def read_register(self, address):
        """Read command and status registers."""
        data = [self.R_REGISTER + address, self.NOP]
        return self.spi.xfer2(data)

    def read_register_multibyte(self, address, length):
        """Read registers RX_ADDR_P0, RX_ADDR_P1 and TX_ADDR."""
        data = [self.R_REGISTER + address] + ([self.NOP] * (length))
        return self.spi.xfer2(data)

    def write_register(self, address, value):
        """Write command and status registers."""
        data = [self.W_REGISTER + address, value]
        return self.spi.xfer2(data)[0]

    def write_register_multibyte(self, address, values):
        """Write registers RX_ADDR_P0, RX_ADDR_P1 and TX_ADDR."""
        data = [self.W_REGISTER + address] + values
        return self.spi.xfer2(data)[0]

    def read_payload(self, length):
        """Read RX-payload: 1 - 32 bytes. A read operation
            always starts at byte 0. Payload is deleted from
            FIFO after it is read. Used in RX mode."""
        data = [self.R_RX_PAYLOAD] + ([self.NOP] * (length))
        return self.spi.xfer2(data)[0]

    def write_payload(self, values):
        """Write TX-payload: 1 - 32 bytes. A write operation
            always starts at byte 0 used in TX payload."""
        data = [self.W_TX_PAYLOAD] + values
        return self.spi.xfer2(data)[0]

    def flush_tx(self):
        """Flush TX FIFO, used in TX mode."""
        return self.spi.xfer2( [self.FLUSH_TX])[0]

    def flush_rx(self):
        """Flush RX FIFO, used in RX mode
            Should not be executed during transmission of
            acknowledge, that is, acknowledge package will
            not be completed."""
        return self.spi.xfer2( [self.FLUSH_RX])[0]

    def reuse_payload(self):
        """Used for a PTX device
            Reuse last transmitted payload.
            TX payload reuse is active until
            W_TX_PAYLOAD or FLUSH TX is executed. TX
            payload reuse must not be activated or deacti-
            vated during package transmission."""
        return self.spi.xfer2( [self.REUSE_TX_PL])[0]

    def read_payload_length(self):
        """Read RX payload width for the top
            R_RX_PAYLOAD in the RX FIFO.
            Note: Flush RX FIFO if the read value is larger
            than 32 bytes"""
        return self.spi.xfer2( [self.R_RX_PL_WID])

    def write_payload_pipe(self, pipe, values):
        """Used in RX mode.
            Write Payload to be transmitted together with
            ACK packet on PIPE pipe. (pipe valid in the
            range from 0 to 5). Maximum three ACK
            packet payloads can be pending. Payloads with
            same pipe are handled using first in - first out
            principle. Write payload: 1- 32 bytes. A write
            operation always starts at byte 0."""
        data = [self.W_ACK_PAYLOAD + pipe] + values
        return self.spi.xfer2(data)[0]

    def write_payload_noack(self, values):
        """Used in TX mode. Disables AUTOACK on this
            specific packet."""
        data = [self.W_TX_PAYLOAD_NOACK] + values
        return self.spi.xfer2(data)[0]

    def read_status(self):
        """Read the STATUS register without performing
            any operation."""
        return self.spi.xfer2( [self.NOP])[0]

    def enable(self):
        """Enable device, send or receive packets depending on PRIM_RX."""
        gpio.output( self.CE, gpio.HIGH)

    def disable(self):
        """Disable device, will enter Standby-I mode."""
        gpio.output( self.CE, gpio.LOW)

    def test(self):
        print("Turn on:")
        print(bin( self.write_register(self.CONFIG, 0b00001010) ))
        print("Write payload")
        print(bin( self.write_payload( [42, 43, 44, 56] ) ))
        print("Enable device")
        self.enable()
        time.sleep(0.0001)
        self.disable()
        print("Status after disabling:")
        print(bin( self.read_status() ))

File: nRF24L01p.c
/* 
 *  Code for the PIC-nRF24L01+ communication.
 *
 *  Compile with: xc8 --chip=16f876A main.c
 *  Load with: #pk2cmd -P PIC16f876A [-X] -M -F main.hex
 *
 *  Author: Simone Pilon <wertyseek@gmail.com>
 */
// vim plugin header: ------------------
// Set compiler:
//!compiler=xc8
// Set device name:
//!device=16F876A
// -------------------------------------

#include <stdio.h>
#include <stdlib.h>

// CONFIGURATION WORD
#pragma config FOSC = HS        // Oscillator Selection bits (HS oscillator)
#pragma config WDTE = OFF       // Watchdog Timer Enable bit (WDT enabled)
#pragma config PWRTE = OFF      // Power-up Timer Enable bit (PWRT disabled)
#pragma config BOREN = ON       // Brown-out Reset Enable bit (BOR enabled)
#pragma config LVP = OFF        // Low-Voltage (Single-Supply) In-Circuit Serial Programming Enable bit (RB3 is digital I/O, HV on MCLR must be used for programming)
#pragma config CPD = OFF        // Data EEPROM Memory Code Protection bit (Data EEPROM code protection off)
#pragma config WRT = OFF        // Flash Program Memory Write Enable bits (Write protection off; all program memory may be written to by EECON control)
#pragma config CP = OFF         // Flash Program Memory Code Protection bit (Code protection off)
#pragma config DEBUG = OFF      // RB6 and 7 are general IO

// #pragma config statements should precede project file includes.
// Use project enums instead of #define for ON and OFF.

#define _XTAL_FREQ 20000000      // Osc. freq. in Hz

#include <xc.h>						// Macros and registers definitions

/*
 * Electrical connections:
 * 	RC0 -> CSN
 * 	RC5 -> MOSI
 * 	RC4 <- MISO
 * 	RC3 -> SCK
 * 	RB0 <- IRQ
 * 	RB1 -> CE
 *
 * 	RB2 -> test clock
 * 	RB3 -> test status
 *
 * Comments:
 *
 * nRF24L01+ usage:
 * 	SPI communication is handled by the SPI interrupt routine.
 * 	1. Load the bytes to be transmitted in spi_buffer (spi_buffer[0] 
 * 		is transmitted last, exclude the first command byte)
 * 		and set spi_counter to the number of bytes to transmit.
 * 	2. Load the command byte (the first) in SSPBUF register 
 * 	3. The rest of the communication is handled automatically, avoid writing
 * 		to spi_counter until spi_ready is 1 (communication is complete).
 */

// Project defines:
// Instruction Mnemonics (nRF24L01+)
#define RF_R_REGISTER 0x00
#define RF_W_REGISTER 0x20
#define RF_R_RX_PAYLOAD 0x61
#define RF_W_TX_PAYLOAD 0xA0
#define RF_FLUSH_TX 0xE1
#define RF_FLUSH_RX 0xE2
#define RF_REUSE_TX_PL 0xE3
#define RF_R_RX_PL_WID 0x60
#define RF_W_ACK_PAYLOAD 0xA8
#define RF_W_TX_PAYLOAD_NOACK 0xB0
#define RF_NOP 0xFF

// Register addresses (nRF24L01+)
#define RF_CONFIG 0x00
#define RF_EN_AA 0x01
#define RF_EN_RXADDR 0x02
#define RF_SETUP_AW 0x03
#define RF_SETUP_RETR 0x04
#define RF_RF_CH 0x05
#define RF_RF_SETUP 0x06
#define RF_STATUS 0x07
#define RF_OBSERVE_TX 0x08
#define RF_RPD 0x09
#define RF_RX_ADDR_P0 0x0A
#define RF_RX_ADDR_P1 0x0B
#define RF_RX_ADDR_P2 0x0C
#define RF_RX_ADDR_P3 0x0D
#define RF_RX_ADDR_P4 0x0E
#define RF_RX_ADDR_P5 0x0F
#define RF_TX_ADDR 0x10
#define RF_RX_PW_P0 0x11
#define RF_RX_PW_P1 0x12
#define RF_RX_PW_P2 0x13
#define RF_RX_PW_P3 0x14
#define RF_RX_PW_P4 0x15
#define RF_RX_PW_P5 0x16
#define RF_FIFO_STATUS 0x17
#define RF_DYNPD 0x1C
#define RF_FEATURE 0x1D

unsigned char spi_buffer[33];		// Bytes queued for transmission
unsigned char spi_counter = 0;	// Number of bytes to transmit
unsigned char spi_ready = 1;		// SPI ready flag

// Main
void main() {
	OPTION_REG =0b10111111;		// RB0 int active low
	// Enable SPI module:
	TRISC = 		0b11010110;		// SDO and SCK outputs, RC0 output
	RC0 = 0b1;
	SSPSTAT =	0b01000000;		// Sampling at middle, trans at high to low CLK transition
	SSPCON =		0b00100001;		// SPI enabled, master with Freq=Fosc/16
	// Interrupts:
	PIE1 = 		0b00001000;		// SPI interrupt enabled
	PIR2 =		0b00000000;		// Clear all peripherals interrupts
	INTCON =		0b11010000;		// Peripheral, RB0 interrupts enabled, flags cleared
	// GPPs
	TRISB =		0b11110001;		// RB1 output
	RB1 = 0b0;
	RB2 = 0b0;
	RB3 = 0b0;

	// Test routine
	__delay_ms(500);
	
	// SPI transaction:
	spi_buffer[0] = 0b00001010;			// Turn RFchip on
	RC0 = 0b0;									// CSN active
	spi_counter = 1;							// Data bytes
	spi_ready = 0;								// Manually clear
	SSPBUF = RF_W_REGISTER + RF_CONFIG;	// Write config register
	while (spi_ready == 0) {				// Wait
		__delay_us(1);
	}
	
	// Send test signal
	RB3 = spi_buffer[1]%2;
	RB2 = 0b1;
	__delay_ms(200);
	RB2 = 0b1;
	
	while (1) {
		// SPI transaction:
		// write payload
		spi_buffer[0] = 56;
		spi_buffer[1] = 44;
		spi_buffer[2] = 43;
		spi_buffer[3] = 42;
		RC0 = 0b0;									// CSN active
		spi_counter = 4;							// Data bytes
		spi_ready = 0;								// Manually clear
		SSPBUF = RF_W_TX_PAYLOAD;				// Write config register
		while (spi_ready == 0) {				// Wait
			__delay_us(1);
		}

		// Send test signal
		RB3 = spi_buffer[4]%2;
		RB2 = 0b1;
		__delay_ms(200);
		RB2 = 0b1;
	}

}

// Interrupt service routine
void interrupt ISR() {
	if (INTF) {			// External interrupt
		INTF = 0b0;
	}
	if (SSPIF) {		// SPI interrupt
		SSPIF = 0b0;
		
		spi_buffer[spi_counter] = SSPBUF;	// Store received value
		if (spi_counter > 0) {
			spi_counter--;
			SSPBUF = spi_buffer[spi_counter];	// Transmit next value
		}
		else {
			// Conclude transaction
			spi_ready = 1;
			RC0 = 0b1;
		}
	}
}
